drag_and_drop_v2/drag_and_drop_v2.py

```diff
@@ -306,23 +306,41 @@ def drop_item(self, item_attempt, suffix=''):
 
     @XBlock.json_handler
     def do_attempt(self, data, suffix=''):
+        ngettext = self._get_ngettext()
         self._validate_attempt()
 
         self.num_attempts += 1
         self._mark_complete_and_publish_grade()
 
-        __, placed, correct = self._get_item_raw_stats()
+        required, placed, correct = self._get_item_raw_stats()
         placed_ids, correct_ids = set(placed), set(correct)
+        missing_ids = set(required) - set(placed)
         misplaced_ids = placed_ids - correct_ids
 
-        feedback_msgs = [_('Correctly placed {correct_count} items, misplaced {misplaced_count} items.').format(
-            correct_count=len(correct_ids),
-            misplaced_count=len(misplaced_ids)
-        )]
-
-        if misplaced_ids:
+        correct_count, misplaced_count, missing_count = len(correct_ids), len(misplaced_ids), len(missing_ids)
+
+        feedback_msgs = [
+            ngettext(
+                'Correctly placed {correct_count} item.',
+                'Correctly placed {correct_count} items.',
+                correct_count
+            ).format(correct_count=correct_count),
+            ngettext(
+                'Misplaced {misplaced_count} item.',
+                'Misplaced {misplaced_count} items.',
+                misplaced_count
+            ).format(misplaced_count=misplaced_count),
+            ngettext(
+                'Not placed {missing_count} required item.',
+                'Not placed {missing_count} required items.',
+                missing_count
+            ).format(missing_count=missing_count)
+        ]
+
+        if misplaced_ids and self.attemps_remain:
             feedback_msgs.append(_('Misplaced items were returned to item bank.'))
-        else:
+
+        if not misplaced_ids and not missing_ids:
```

Comment:
I was about to suggest to replace this with `if self._is_finished()` to make it clearer, but then I realized that `_is_finished()` is currently a bit broken in assessment mode, because it does not check decoy items. `_is_finished()` should only return true if all non-decoy items are in their correct positions and all decoy items are in the bank.